src/Microsoft.AspNetCore.Mvc.Razor/Internal/CompilerCache.cs

```diff
@@ -126,19 +136,19 @@ public CompilerCache(IFileProvider fileProvider)
 
                 cacheEntryOptions = new MemoryCacheEntryOptions();
 
-                compilerCacheContext = cacheContextFactory(normalizedPath);
-                cacheEntryOptions.ExpirationTokens.Add(_fileProvider.Watch(compilerCacheContext.ProjectItem.Path));
-                if (!compilerCacheContext.ProjectItem.Exists)
+                cacheEntryOptions.ExpirationTokens.Add(_fileProvider.Watch(normalizedPath));
+                var projectItem = _templateEngine.Project.GetItem(normalizedPath);
```

Comment:
I don't feel like this should be coupled to project items and the template engine in particular. The template engine is the server, what's the abstraction for a client?